### PR TITLE
Add BlockBreakBlockEvent

### PR DESCRIPTION
**Pull request by Machine-Maker**
So not really sure what to call this event, but there wasn't a way to affect the drops when water or a piston breaks a block. So this adds that. Kind of an odd name I know, but if you've got a better one, suggest away please.

### CODE REVIEW

**Comment by ExcessiveAmountsOfZombies** [comment]
EnvironmentalBlockDropEvent? ExternalBlockDropEvent?

**Comment by electronicboy** [comment]
I got an idea.
Kassssshhhhhh, bring ur wisdoms

**Comment by Geolykt** [comment]
~~BlockReplacementEvent~~ (this is a shit suggestion) or a variation thereof (such as BlockReplacedEvent or ReplacedBlockDropEvent) as the block is destroyed so another block (in this case either liquid or piston head) is set on where the old block once was

**Comment by aurorasmiles** [comment]
BlockBreakByEnvironmentItemDropEvent :D

**Comment by stefvanschie** [comment]
I like ExternalBlockDropEvent as suggested by ExcessiveAmountOfZombies.

**Comment by yannicklamprecht** [comment]
DropItemByEnvironmentalBlockBreakEvent

**Comment by Phoenix616** [comment]
Imo. the naming difficulty exposes a completely different issue: The event only gets called on a block drop, not on the block break making it impossible to cancel that event as well as getting the drop. Imo. that should be combined to one event which both handles the actual block breaking as well as the drop catching. Edit: Granted, the implementation might not be possible that way and the Player block breaking also consists of two events (which I don't really like) :S

That could also make the naming easier:

- `BlockBlockBreakEvent`
- `EnvironmentBlockBreakEvent`
- stranger but might sound better: `BlockBreakBlockEvent`

**Comment by Machine-Maker** [comment]
K, rebased (somehow I forgot this one after md_5's nms repackage), and I went with `BlockBreakBlockEvent`

**Review by electronicboy — APPROVED** [review]

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

File: patches/server/0725-Add-BlockBreakBlockEvent.patch [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+@@ -331,6 +331,19 @@ public class Block extends BlockBehaviour implements ItemLike {
+         }
+ 
+     }
++    // Paper start
++    public static boolean dropItemsWithEvent(BlockState blockState, LevelAccessor generatorAccess, BlockPos blockPosition, BlockPos source, @Nullable BlockEntity tileEntity) {
++        if (generatorAccess instanceof ServerLevel) {
++            io.papermc.paper.event.block.BlockBreakBlockEvent event = new io.papermc.paper.event.block.BlockBreakBlockEvent(org.bukkit.craftbukkit.block.CraftBlock.at(generatorAccess, blockPosition), org.bukkit.craftbukkit.block.CraftBlock.at(generatorAccess, source), Block.getDrops(blockState, generatorAccess.getMinecraftWorld(), blockPosition, tileEntity).stream().map(org.bukkit.craftbukkit.inventory.CraftItemStack::asBukkitCopy).collect(java.util.stream.Collectors.toList()));
++            event.callEvent();
++            event.getDrops().stream().map(org.bukkit.craftbukkit.inventory.CraftItemStack::asNMSCopy).forEach((itemstack) -> {
```
Comment (Proximyst) [inline patches/server/0725-Add-BlockBreakBlockEvent.patch]: Feels like a small for loop would be nicer here, as I reckon this is called for every block break?

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: patches/server/0725-Add-BlockBreakBlockEvent.patch (hunk 1 above)
Comment (Machine-Maker) [reply patches/server/0725-Add-BlockBreakBlockEvent.patch]: It's not, just when other blocks cause a block to be broken, like a piston extension or water flowing. BUT it doesnt hurt to replace the streams anyways.

**Comment by qixils** [comment]
a `BlockDropResourcesEvent` that gets called for every `#dropResources` call would be nice for modifying i.e. drops of decaying leaves which aren't explicitly caused by another block. this event could extend that as well? probably out of scope for this PR, just figured i'd note it here since it's pretty similar, and it's a feature i wanted so much i [implemented it myself](https://github.com/ProjectEdenGG/Parchment/blob/master/patches/server/0011-Add-BlockDropResourcesEvent.patch) not too differently from this PR :P _(yes, it uses streams, i'll fix it eventually)_

**Comment by Brokkonaut** [comment]
Is there anything still blocking this PR? This feature would be really helpful for the plugin I am currently creating...

File: patches/server/0726-Add-BlockBreakBlockEvent.patch [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
++++ b/src/main/java/net/minecraft/world/level/block/Block.java
+@@ -331,6 +331,23 @@ public class Block extends BlockBehaviour implements ItemLike {
+         }
+ 
+     }
++    // Paper start
++    public static boolean dropResources(BlockState blockState, LevelAccessor generatorAccess, BlockPos blockPosition, BlockPos source, @Nullable BlockEntity tileEntity) {
++        if (generatorAccess instanceof ServerLevel) {
++            List<org.bukkit.inventory.ItemStack> items = com.google.common.collect.Lists.newArrayList();
++            for (net.minecraft.world.item.ItemStack drop : net.minecraft.world.level.block.Block.getDrops(blockState, generatorAccess.getMinecraftWorld(), blockPosition, tileEntity)) {
```
Comment (Proximyst) [inline patches/server/0726-Add-BlockBreakBlockEvent.patch]: Fully qualified name here could be replaced with `var` but it's no biggie

**Review by Proximyst — APPROVED** [review]
LGTM :)

**Pull request merged**